File: ios/Jazzify/EarTraining/Engine/EarTrainingChordQuiz.swift
import Foundation

/// Web `pickNextQuizIndex` / `isQuizClear`（[src/utils/earTrainingChordQuiz.ts] と同じ契約）。
enum EarTrainingChordQuiz {
    struct Question: Identifiable, Sendable {
        let id: String
        let orderIndex: Int
        let measureNumber: Int?
        let keyFifths: Int?
        let chords: [EarTrainingPhraseChordDetail]
    }

    enum QuestionOrder {
        case random
        case sequential

        init(sequential: Bool) {
            self = sequential ? .sequential : .random
        }
    }

    static func pickNextQuizIndex<Item>(
        items: [Item],
        order: QuestionOrder,
        prevIndex: Int?,
        rand: () -> Double
    ) -> Int {
        let n = items.count
        if n <= 0 {
            return 0
        }
        if n == 1 {
            return 0
        }
        if order == .sequential {
            return ((prevIndex ?? -1) + 1) % n
        }
        var next = Int(floor(rand() * Double(n)))
        if let prevIndex, next == prevIndex {
            var guardCount = 0
            while next == prevIndex && guardCount < 32 {
                next = Int(floor(rand() * Double(n)))
                guardCount += 1
            }
            if next == prevIndex {
                next = (prevIndex + 1) % n
            }
        }
        return next
    }

    static func isQuizClear(correct: Int, required: Int) -> Bool {
        correct >= max(0, required)
    }

    static func buildQuestions(stage: EarTrainingStageDetail) -> [Question] {
        let itemQuestions = buildQuestions(
            items: stage.sortedChordQuizItems(),
            syntheticPhraseId: stage.id
        )
        if !itemQuestions.isEmpty {
            return itemQuestions
        }
        return buildQuestions(phrases: stage.sortedPhrases())
    }

    static func activeChord(in question: Question?, completedChordIds: Set<UUID>?) -> EarTrainingPhraseChordDetail? {
        guard let question else { return nil }
        return question.chords.first { chord in
            chordHasPlayableVoicing(chord) && (completedChordIds?.contains(chord.id) ?? false) == false
        }
    }

    static func isQuestionCompleted(_ question: Question?, completedChordIds: Set<UUID>) -> Bool {
        guard let question, !question.chords.isEmpty else { return false }
        return question.chords.allSatisfy { chord in
            !chordHasPlayableVoicing(chord) || completedChordIds.contains(chord.id)
        }
    }

    private struct MutableGroup {
        let id: String
        var orderIndex: Int
        let measureNumber: Int?
        let keyFifths: Int?
        var chords: [EarTrainingPhraseChordDetail]
    }

    private static func chordHasPlayableVoicing(_ chord: EarTrainingPhraseChordDetail) -> Bool {
        !(chord.voicing ?? []).isEmpty
    }

    private static func normalizedPositiveInteger(_ value: Int?) -> Int? {
        guard let value, value >= 1 else { return nil }
        return value
    }

    private static func sortChordsForQuestion(
        _ lhs: EarTrainingPhraseChordDetail,
        _ rhs: EarTrainingPhraseChordDetail
    ) -> Bool {
        let lm = lhs.measureNumber ?? Int.max
        let rm = rhs.measureNumber ?? Int.max
        if lm != rm { return lm < rm }
        let lb = lhs.beatOffset ?? Double.greatestFiniteMagnitude
        let rb = rhs.beatOffset ?? Double.greatestFiniteMagnitude
        if lb != rb { return lb < rb }
        let ls = lhs.startTimeSec ?? Double.greatestFiniteMagnitude
        let rs = rhs.startTimeSec ?? Double.greatestFiniteMagnitude
        if ls != rs { return ls < rs }
        return lhs.orderIndex < rhs.orderIndex
    }

    private static func question(from group: MutableGroup) -> Question {
        Question(
            id: group.id,
            orderIndex: group.orderIndex,
            measureNumber: group.measureNumber,
            keyFifths: group.keyFifths,
            chords: group.chords.sorted(by: sortChordsForQuestion)
        )
    }

    private static func chord(from item: EarTrainingChordQuizItem, phraseId: UUID) -> EarTrainingPhraseChordDetail {
        EarTrainingPhraseChordDetail(
            id: item.id,
            phraseId: phraseId,
            orderIndex: item.orderIndex,
            chordName: item.chordName,
            measureNumber: item.measureNumber,
            beatOffset: item.beatOffset,
            durationBeats: item.durationBeats,
            startTimeSec: nil,
            endTimeSec: nil,
            voicing: item.voicing,
            voicingStaves: item.voicingStaves,
            quote: nil
        )
    }

    private static func buildQuestions(
        items: [EarTrainingChordQuizItem],
        syntheticPhraseId: UUID
    ) -> [Question] {
        var groups: [MutableGroup] = []
        var groupIndexByMeasure: [Int: Int] = [:]
        let sortedItems = items
            .filter { !$0.voicing.isEmpty }
            .sorted {
                let lhs = chord(from: $0, phraseId: syntheticPhraseId)
                let rhs = chord(from: $1, phraseId: syntheticPhraseId)
                return sortChordsForQuestion(lhs, rhs)
            }

        for item in sortedItems {
            guard let measureNumber = normalizedPositiveInteger(item.measureNumber) else {
                let phraseId = "chord-quiz-item-\(item.id.uuidString)"
                groups.append(
                    MutableGroup(
                        id: phraseId,
                        orderIndex: item.orderIndex,
                        measureNumber: nil,
                        keyFifths: nil,
                        chords: [chord(from: item, phraseId: syntheticPhraseId)]
                    )
                )
                continue
            }

            if let index = groupIndexByMeasure[measureNumber] {
                groups[index].orderIndex = min(groups[index].orderIndex, item.orderIndex)
                groups[index].chords.append(chord(from: item, phraseId: syntheticPhraseId))
            } else {
                groupIndexByMeasure[measureNumber] = groups.count
                groups.append(
                    MutableGroup(
                        id: "chord-quiz-measure-\(measureNumber)",
                        orderIndex: item.orderIndex,
                        measureNumber: measureNumber,
                        keyFifths: nil,
                        chords: [chord(from: item, phraseId: syntheticPhraseId)]
                    )
                )
            }
        }

        return groups
            .map { question(from: $0) }
            .sorted { $0.orderIndex < $1.orderIndex }
    }

    private static func buildQuestions(phrases: [EarTrainingPhraseDetail]) -> [Question] {
        var groups: [MutableGroup] = []
        for phrase in phrases {
            var groupIndexByMeasure: [Int: Int] = [:]
            let chords = (phrase.chords ?? [])
                .filter(chordHasPlayableVoicing)
                .sorted(by: sortChordsForQuestion)

            for chord in chords {
                guard let measureNumber = normalizedPositiveInteger(chord.measureNumber) else {
                    groups.append(
                        MutableGroup(
                            id: "\(phrase.id.uuidString)-chord-\(chord.id.uuidString)",
                            orderIndex: groups.count,
                            measureNumber: nil,
                            keyFifths: phrase.keyFifths,
                            chords: [chord]
                        )
                    )
                    continue
                }

                if let index = groupIndexByMeasure[measureNumber] {
                    groups[index].chords.append(chord)
                } else {
                    groupIndexByMeasure[measureNumber] = groups.count
                    groups.append(
                        MutableGroup(
                            id: "\(phrase.id.uuidString)-measure-\(measureNumber)",
                            orderIndex: groups.count,
                            measureNumber: measureNumber,
                            keyFifths: phrase.keyFifths,
                            chords: [chord]
                        )
                    )
                }
            }
        }
        return groups.map { question(from: $0) }
    }
}

File: src/utils/earTrainingChordQuiz.ts
import type {
  EarTrainingChordQuizItem,
  EarTrainingPhrase,
  EarTrainingPhraseChord,
  EarTrainingStage,
} from '@/types';

/**
 * chord_quiz: 次の出題インデックス（読み取りやすさのため `items[].order_index` は使わず 0..n-1）。
 * `random`: 問題が2つ以上あるとき `prevIndex` と別の問題を選ぶ。
 */
export const pickNextQuizIndex = <T extends { order_index?: number | null }>(
  items: readonly T[],
  order: 'random' | 'sequential',
  prevIndex: number | null,
  rand: () => number,
): number => {
  const n = items.length;
  if (n <= 0) {
    return 0;
  }
  if (n === 1) {
    return 0;
  }
  if (order === 'sequential') {
    return ((prevIndex ?? -1) + 1) % n;
  }
  let next = Math.floor(rand() * n);
  if (prevIndex !== null && next === prevIndex) {
    let guard = 0;
    while (next === prevIndex && guard < 32) {
      next = Math.floor(rand() * n);
      guard += 1;
    }
    if (next === prevIndex) {
      next = (prevIndex + 1) % n;
    }
  }
  return next;
};

export const isQuizClear = (correct: number, required: number): boolean => (
  correct >= Math.max(0, required)
);

export interface EarTrainingChordQuizQuestion {
  id: string;
  order_index: number;
  measure_number: number | null;
  key_fifths?: number | null;
  chords: readonly EarTrainingPhraseChord[];
}

interface MutableQuestionGroup {
  id: string;
  orderIndex: number;
  measureNumber: number | null;
  keyFifths?: number | null;
  chords: EarTrainingPhraseChord[];
}

const hasPlayableVoicing = (
  chord: Pick<EarTrainingPhraseChord, 'voicing'>,
): boolean => (chord.voicing?.length ?? 0) > 0;

const normalizedPositiveInteger = (value: number | null | undefined): number | null => {
  if (value === null || value === undefined || !Number.isFinite(value)) {
    return null;
  }
  const normalized = Math.trunc(value);
  return normalized >= 1 ? normalized : null;
};

const sortChordsForQuizQuestion = <T extends {
  order_index: number;
  measure_number?: number | null;
  beat_offset?: number | null;
  start_time_sec?: number | null;
}>(a: T, b: T): number => {
  const measureA = a.measure_number ?? Number.POSITIVE_INFINITY;
  const measureB = b.measure_number ?? Number.POSITIVE_INFINITY;
  if (measureA !== measureB) {
    return measureA - measureB;
  }
  const beatA = a.beat_offset ?? Number.POSITIVE_INFINITY;
  const beatB = b.beat_offset ?? Number.POSITIVE_INFINITY;
  if (beatA !== beatB) {
    return beatA - beatB;
  }
  const startA = a.start_time_sec ?? Number.POSITIVE_INFINITY;
  const startB = b.start_time_sec ?? Number.POSITIVE_INFINITY;
  if (startA !== startB) {
    return startA - startB;
  }
  return a.order_index - b.order_index;
};

const questionFromGroup = (group: MutableQuestionGroup): EarTrainingChordQuizQuestion => ({
  id: group.id,
  order_index: group.orderIndex,
  measure_number: group.measureNumber,
  key_fifths: group.keyFifths,
  chords: group.chords.slice().sort(sortChordsForQuizQuestion),
});

const quizItemToPhraseChord = (
  item: EarTrainingChordQuizItem,
  phraseId: string,
): EarTrainingPhraseChord => ({
  id: item.id,
  phrase_id: phraseId,
  order_index: item.order_index,
  chord_name: item.chord_name,
  measure_number: item.measure_number ?? null,
  beat_offset: item.beat_offset ?? null,
  duration_beats: item.duration_beats ?? null,
  voicing: item.voicing,
  voicing_staves: item.voicing_staves,
});

const buildEarTrainingChordQuizQuestionsFromItems = (
  items: readonly EarTrainingChordQuizItem[],
): EarTrainingChordQuizQuestion[] => {
  const groups: MutableQuestionGroup[] = [];
  const groupByMeasure = new Map<number, MutableQuestionGroup>();

  items
    .filter(item => item.voicing.length > 0)
    .slice()
    .sort(sortChordsForQuizQuestion)
    .forEach(item => {
      const measureNumber = normalizedPositiveInteger(item.measure_number);
      if (measureNumber === null) {
        const phraseId = `chord-quiz-item-${item.id}`;
        groups.push({
          id: phraseId,
          orderIndex: item.order_index,
          measureNumber: null,
          chords: [quizItemToPhraseChord(item, phraseId)],
        });
        return;
      }

      let group = groupByMeasure.get(measureNumber);
      if (!group) {
        group = {
          id: `chord-quiz-measure-${measureNumber}`,
          orderIndex: item.order_index,
          measureNumber,
          chords: [],
        };
        groupByMeasure.set(measureNumber, group);
        groups.push(group);
      }
      group.orderIndex = Math.min(group.orderIndex, item.order_index);
      group.chords.push(quizItemToPhraseChord(item, group.id));
    });

  return groups
    .map(questionFromGroup)
    .sort((a, b) => a.order_index - b.order_index);
};

const buildEarTrainingChordQuizQuestionsFromPhrases = (
  phrases: readonly EarTrainingPhrase[],
): EarTrainingChordQuizQuestion[] => {
  const groups: MutableQuestionGroup[] = [];

  phrases
    .slice()
    .sort((a, b) => a.order_index - b.order_index)
    .forEach(phrase => {
      const groupByMeasure = new Map<number, MutableQuestionGroup>();
      const phraseChords = (phrase.chords ?? [])
        .filter(hasPlayableVoicing)
        .slice()
        .sort(sortChordsForQuizQuestion);

      phraseChords.forEach(chord => {
        const measureNumber = normalizedPositiveInteger(chord.measure_number);
        if (measureNumber === null) {
          groups.push({
            id: `${phrase.id}-chord-${chord.id}`,
            orderIndex: groups.length,
            measureNumber: null,
            keyFifths: phrase.key_fifths,
            chords: [chord],
          });
          return;
        }

        let group = groupByMeasure.get(measureNumber);
        if (!group) {
          group = {
            id: `${phrase.id}-measure-${measureNumber}`,
            orderIndex: groups.length,
            measureNumber,
            keyFifths: phrase.key_fifths,
            chords: [],
          };
          groupByMeasure.set(measureNumber, group);
          groups.push(group);
        }
        group.chords.push(chord);
      });
    });

  return groups.map(questionFromGroup);
};

export const buildEarTrainingChordQuizQuestions = (
  stage: Pick<EarTrainingStage, 'chord_quiz_items' | 'phrases'>,
): EarTrainingChordQuizQuestion[] => {
  const itemQuestions = buildEarTrainingChordQuizQuestionsFromItems(stage.chord_quiz_items ?? []);
  if (itemQuestions.length > 0) {
    return itemQuestions;
  }
  return buildEarTrainingChordQuizQuestionsFromPhrases(stage.phrases ?? []);
};

export const getActiveChordInQuizQuestion = (
  question: EarTrainingChordQuizQuestion | null | undefined,
  completedChordIds: ReadonlySet<string> | null | undefined,
): EarTrainingPhraseChord | null => {
  const chords = question?.chords ?? [];
  if (chords.length === 0) {
    return null;
  }
  return chords.find(
    chord => hasPlayableVoicing(chord) && completedChordIds?.has(chord.id) !== true,
  ) ?? null;
};

export const isChordQuizQuestionCompleted = (
  question: EarTrainingChordQuizQuestion | null | undefined,
  completedChordIds: ReadonlySet<string>,
): boolean => {
  const chords = question?.chords ?? [];
  return chords.length > 0 && chords.every(chord => !hasPlayableVoicing(chord) || completedChordIds.has(chord.id));
};
